graphql-dgs-spring-webmvc/src/main/kotlin/com/netflix/graphql/dgs/mvc/DgsRestController.kt

```diff
@@ -222,3 +243,135 @@ open class DgsRestController(
+    fun assert(headers: HttpHeaders)
+
+    open class DgsGraphqlRequestHeaderValidationException(s: String) : IllegalArgumentException(s)
+
+    open class DgsGraphqlRequestContentTypePredicateException(s: String) : DgsGraphqlRequestHeaderValidationException(s)
+
+    open class DgsGraphQLRequestHeaderRuleException(s: String) : DgsGraphqlRequestHeaderValidationException(s)
+}
+
+@Order
+fun interface DgsGraphQLRequestContentTypePredicate {
+    fun accept(contentType: MediaType?): Boolean
+
+    companion object {
+        private val ALLOW_ALL_CONTENT_TYPE_PREDICATE = DgsGraphQLRequestContentTypePredicate { _ -> true }
+
+        private val STRICT_GRAPHQL_CONTENT_TYPES =
+            listOf(MediaType.APPLICATION_JSON, MediaType.MULTIPART_FORM_DATA, GraphQLMediaTypes.GRAPHQL_MEDIA_TYPE)
+
+        val STRICT_GRAPHQL_CONTENT_TYPES_PREDICATE = DgsGraphQLRequestContentTypePredicate { mediaType ->
+            mediaType != null && STRICT_GRAPHQL_CONTENT_TYPES.find { it.isCompatibleWith(mediaType) } != null
+        }
+        val RECOMMENDED_GRAPHQL_CONTENT_TYPE_PREDICATES = listOf(STRICT_GRAPHQL_CONTENT_TYPES_PREDICATE)
+    }
+}
+
+@Order
+fun interface DgsGraphQLRequestHeaderValidationRule {
+    @kotlin.jvm.Throws(DgsGraphQLRequestHeadersValidator.DgsGraphqlRequestContentTypePredicateException::class)
+    fun assert(headers: HttpHeaders)
+}
+
+class DefaultDgsGraphQLRequestHeaderValidator(
+    private val contentTypePredicates: List<DgsGraphQLRequestContentTypePredicate> = listOf(STRICT_GRAPHQL_CONTENT_TYPES_PREDICATE),
+    private val validationRules: List<DgsGraphQLRequestHeaderValidationRule> = DgsGraphQLRequestHeadersValidator.RECOMMENDED_GRAPHQL_REQUEST_HEADERS_VALIDATOR,
+) : DgsGraphQLRequestHeadersValidator {
+    override fun assert(headers: HttpHeaders) {
+        if (contentTypePredicates.isNotEmpty()) {
+            contentTypePredicates.find { it.accept(headers.contentType) }
+                ?: throw DgsGraphQLRequestHeadersValidator.DgsGraphqlRequestContentTypePredicateException("Unsupported Content-Type ${headers.contentType}")
+        }
+        validationRules.forEach { it.assert(headers) }
+    }
+}
+
+class DgsGraphQLCSRFRequestHeaderValidationRule : DgsGraphQLRequestHeaderValidationRule {
+    companion object {
+        // CSRF Prevention Request Headers
+        @Suppress("MemberVisibilityCanBePrivate")
+        const val HEADER_X_APOLLO_OPERATION_NAME = "x-apollo-operation-name"
+        @Suppress("MemberVisibilityCanBePrivate")
+        const val HEADER_APOLLO_REQUIRE_PREFLIGHT = "apollo-require-preflight"
+        @Suppress("MemberVisibilityCanBePrivate")
+        const val HEADER_GRAPHQL_REQUIRE_PREFLIGHT = "graphql-require-preflight"
+
+        /**
+         * Our recommended set of CSRF prevention headers. Operations that do not provide a content-type such as `application/json`
+         * (in practice, this means GET operations) must include at least one of these headers.
```

Comment:
This comment (possibly borrowed from Apollo Server :) ) is outdated given that multipart/form-data qualifies here too.

Reply:
Thanks @glasser , will update the docs. The CSRF check is _heavily inspired_ by the Apollo Server implementation, and the comment was blatantly copied from it.